### PR TITLE
Usergroups

### PR DESCRIPTION
**Pull request by pavkahanov**

### CODE REVIEW

File: contracts/ZeroOne/UserGroups/UserGroups.sol [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  }
+
+  constructor() public {}
+
+  /**
+   * @notice gets group data
+   * @param _id group id
+   * @return group
+  */
+  function getGroup(
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/UserGroups.sol]: getUserGroup

File: contracts/ZeroOne/UserGroups/UserGroups.sol [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+  {
+    return groups.list.length;
+  }
+
+  /**
+   * @notice adds new group to list
+   * @param _group group
+   * @return id
+  */
+  function addGroup(
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/UserGroups.sol]: addUserGroup

File: contracts/ZeroOne/UserGroups/UserGroups.sol [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    UserGroup.Group memory _group
+  )
+    public
+    virtual
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/UserGroups.sol]: What are you going to override?

File: contracts/ZeroOne/UserGroups/lib/UserGroup.sol [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+  */
+  function validate(
+    Group memory _group
+  )
+    internal
+    pure
+    returns (bool valid)
+  {
+    return (
+      _group.groupAddress != address(0)
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/lib/UserGroup.sol]: add todo to check that provided address is contract

File: contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    List storage _self,
+    uint _id
+  )
+    internal
+    view
+    returns (bool valid)
+  {
+    return _self.list.length > _id;
+  }
+
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol]: extra line break

File: contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    uint _id
+  )
+    internal
+    view
+    returns (bool valid)
+  {
+    return _self.list.length > _id;
+  }
+  
+}
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol]: add new line at the end of file

File: test/7_UserGroups.spec.js [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  }
+
+  beforeEach(async () => {
+    userGroups = await UserGroups.new({ from })
+  })
+
+  describe('addGroup()', () => {
+    it('should successfully add new user group', async () => {
+      let length = await userGroups.getUserGroupsAmount();
+      assert.strictEqual(length.toNumber(), 0);
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: separate creation test

File: test/7_UserGroups.spec.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      const {name, groupAddress, groupType} = await userGroups.getGroup(0);
+      assert.strictEqual(name, group.name)
+      assert.strictEqual(groupAddress, group.groupAddress)
+      assert.strictEqual(Number(groupType), group.groupType)
+    });
+
+    it('should fail on adding user group with non-unique name', async () => {
+      let error = false; 
+      try {
+        await userGroups.addGroup(group);
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: this line should be before try

File: test/7_UserGroups.spec.js [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+        await userGroups.addGroup(group);
+      } catch {
+        error = true;
+      }
+      assert.strictEqual(error, true);
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: should check erro message

File: test/7_UserGroups.spec.js [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+    it('should fail on adding user group without address', async () => {
+      let error = false; 
+      try {
+        const wrongGroup = Object.create(group);
+        wrongGroup.groupAddress = '0x';
+        await userGroups.addGroup(wrongGroup);
+      } catch {
+        error = true;
+      }
+      assert.strictEqual(error, true);
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: should check erro message

File: test/7_UserGroups.spec.js [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+  describe('events', () => {
+    it('should fire event when added new usergroup', async () => {
+      const tx = await userGroups.addGroup(group);
+      const log = tx.logs.find(element => element.event.match('UserGroupAdded'));
+      const {args: {name}} = log;
+      assert.strictEqual(name, group.name);
+    });
+  });
+})
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: add extra line

File: contracts/ZeroOne/UserGroups/UserGroups.sol [hunk 12]
```diff
@@ -0,0 +1,86 @@
+pragma solidity 0.6.1;
+pragma experimental ABIEncoderV2;
+
+import "./lib/UserGroupsList.sol";
+import "./lib/UserGroup.sol";
+
+/**
+ * @title UserGroups
+ * @dev stores usergroups
```
[the remark this answers is not in the document]
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: fixed

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 1 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: Fixed

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 2 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: fixed

File: contracts/ZeroOne/UserGroups/lib/UserGroup.sol (hunk 4 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/lib/UserGroup.sol]: added

File: contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol (hunk 5 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol]: fixed

File: contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol (hunk 6 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/lib/UserGroupsList.sol]: added

File: test/7_UserGroups.spec.js (hunk 8 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: fixed

File: test/7_UserGroups.spec.js (hunk 7 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: fixed

File: test/7_UserGroups.spec.js (hunk 11 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: added

File: test/7_UserGroups.spec.js (hunk 8 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: fixed

File: test/7_UserGroups.spec.js (hunk 7 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: separated

File: contracts/ZeroOne/UserGroups/lib/UserGroup.sol (hunk 4 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/lib/UserGroup.sol]: added

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 12 above)
[the remark this answers is not in the document]
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: fixed

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 1 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: fixed

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 2 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: fixed

File: test/7_UserGroups.spec.js (hunk 9 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: fixed

File: test/7_UserGroups.spec.js (hunk 10 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: fixed

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 3 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: removed virtual mark

File: contracts/Interfaces/IERC20.sol [hunk 13]
```diff
@@ -0,0 +1,79 @@
+pragma solidity 0.6.1;
```
Comment (masteralx) [inline contracts/Interfaces/IERC20.sol]: move to UserGroups
IERC20.sol

File: contracts/Interfaces/IERC20.sol [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+     * allowance.
+     *
+     * Returns a boolean value indicating whether the operation succeeded.
+     *
+     * Emits a {Transfer} event.
+     */
+    function transferFrom(address sender, address recipient, uint256 amount) external returns (bool);
+
+
+    function owner() external view returns (address);
```
Comment (masteralx) [inline contracts/Interfaces/IERC20.sol]: add docs

File: contracts/Interfaces/IERC20.sol [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+     * Note that `value` may be zero.
+     */
+    event Transfer(address indexed from, address indexed to, uint256 value);
+
+    /**
+     * @dev Emitted when the allowance of a `spender` for an `owner` is set by
+     * a call to {approve}. `value` is the new allowance.
+     */
+    event Approval(address indexed owner, address indexed spender, uint256 value);
+}
```
Comment (masteralx) [inline contracts/Interfaces/IERC20.sol]: add extra line

File: contracts/ZeroOne/UserGroups/UserGroups.sol [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+    @param _groupId group id
+    @return totalSupply
+   */
+  function getTotalSupply(uint _groupId)
+    public
+    view
+    groupExists(_groupId)
+    returns(uint256 totalSupply) 
+  {
+    UserGroup.Group memory userGroup = getUserGroup(_groupId);
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/UserGroups.sol]: check gas use for group[_groupId]

File: contracts/ZeroOne/UserGroups/UserGroups.sol [hunk 17]
```diff
@@ -0,0 +1,121 @@
...
+    public
+    view
+    groupExists(_groupId)
+    returns(address admin)
+  {
+    UserGroup.Group memory userGroup = getUserGroup(_groupId);
+    require(userGroup.groupType == UserGroup.Type.CUSTOM, "This is not custom token group");
+    IERC20 group = IERC20(userGroup.groupAddress);
+    admin = group.owner();
+  }
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/UserGroups.sol]: move it to UserGroup as getAdmin() or admin()

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 16 above)
Comment (masteralx) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: move it to UserGroup as getTotalSupply() or totalSupply()

File: contracts/__mocks__/TokenMock.sol [hunk 18]
```diff
@@ -0,0 +1,23 @@
+pragma solidity 0.6.1;
```
Comment (masteralx) [inline contracts/__mocks__/TokenMock.sol]: add docs

File: test/7_UserGroups.spec.js [hunk 19]
```diff
@@ -0,0 +1,138 @@
...
+        assert.strictEqual(message, getErrorMessage("Name must be unique"));
+      }
+      assert.strictEqual(error, true);
+    });
+
+    it('should fail on adding user group without address', async () => {
+      let error = false; 
+      try {
+        const wrongGroup = Object.create(group);
+        wrongGroup.groupAddress = '0x';
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: try `0x${Array(40).fill(0).join('')}`

File: contracts/Interfaces/IERC20.sol (hunk 13 above)
Comment (pavkahanov) [reply contracts/Interfaces/IERC20.sol]: moved

File: contracts/Interfaces/IERC20.sol (hunk 15 above)
Comment (pavkahanov) [reply contracts/Interfaces/IERC20.sol]: added

File: contracts/__mocks__/TokenMock.sol (hunk 18 above)
Comment (pavkahanov) [reply contracts/__mocks__/TokenMock.sol]: added

File: test/7_UserGroups.spec.js (hunk 19 above)
Comment (pavkahanov) [reply test/7_UserGroups.spec.js]: fixed

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 17 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: Moved

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 16 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: moved, but gas not checked

File: contracts/Interfaces/IERC20.sol (hunk 14 above)
Comment (pavkahanov) [reply contracts/Interfaces/IERC20.sol]: added

File: contracts/ZeroOne/UserGroups/UserGroups.sol (hunk 16 above)
Comment (pavkahanov) [reply contracts/ZeroOne/UserGroups/UserGroups.sol]: > check gas use for group[_groupId]

i think we don't need to do this?

File: contracts/ZeroOne/UserGroups/lib/UserGroup.sol [hunk 20]
```diff
@@ -0,0 +1,70 @@
+pragma solidity 0.6.1;
+// pragma experimental ABIEncoderV2;
```
Comment (masteralx) [inline contracts/ZeroOne/UserGroups/lib/UserGroup.sol]: ?

File: contracts/__mocks__/ERC20Mock.sol [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+      _totalSupply = totalSupply;
+    }
+
+    /**
+     * @dev See {IERC20-totalSupply}.
+     */
+    function totalSupply() public view returns (uint256) {
+        return _totalSupply;
+    }
+}
```
Comment (masteralx) [inline contracts/__mocks__/ERC20Mock.sol]: extra line

File: contracts/__mocks__/TokenMock.sol [hunk 22]
```diff
@@ -0,0 +1,33 @@
...
+    return _totalSupply;
+  }
+  
+  /**
+    @notice method for get admin of tokens
+   */
+  function owner() public view returns(address) {
+    return _owner;
+  }
+}
```
Comment (masteralx) [inline contracts/__mocks__/TokenMock.sol]: extra line

File: contracts/__mocks__/UserGroupMock.sol [hunk 23]
```diff
@@ -0,0 +1,45 @@
+
+
```
Comment (masteralx) [inline contracts/__mocks__/UserGroupMock.sol]: ?

File: test/7_UserGroups.spec.js [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+        groupType: 0,
+        groupAddress: address
+      }
+      await userGroups.addUserGroup(group)
+      try {
+        const admin = await userGroups.getUserGroupAdmin(0)
+      } catch({ message }) {
+        assert.strictEqual(message, getShortErrorMessage('This is not custom token group'));
+      }
+    });
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: assert.strictEqual(error, true);

File: test/7_UserGroups.spec.js [hunk 25]
```diff
@@ -0,0 +1,138 @@
...
+  describe('events', () => {
+    it('should fire event when added new usergroup', async () => {
+      const tx = await userGroups.addUserGroup(group);
+      const log = tx.logs.find(element => element.event.match('UserGroupAdded'));
+      const {args: {name}} = log;
+      assert.strictEqual(name, group.name);
+    });
+  });
+
+});
```
Comment (masteralx) [inline test/7_UserGroups.spec.js]: extra line

File: test/9_UserGroup.spec.js [hunk 26]
```diff
@@ -0,0 +1,90 @@
+const UserGroupMock = artifacts.require('./UserGroupMock.sol');
+const TokenMock = artifacts.require('./TokenMock.sol');
+const ERC20Mock = artifacts.require('./ERC20.sol');
+
+const { getShortErrorMessage } = require('./helpers/get-error-message');
+
+
```
Comment (masteralx) [inline test/9_UserGroup.spec.js]: remove extra line

File: test/9_UserGroup.spec.js [hunk 27]
```diff
@@ -0,0 +1,90 @@
...
+      });
+      userGroup = await UserGroupMock.new(ERC20);
+      
+      try {
+        await userGroup.testGetAdmin();
+      } catch ({ message }) {
+        assert.strictEqual(message, getShortErrorMessage('This is not custom token group'))
+      }
+    });
+
```
Comment (masteralx) [inline test/9_UserGroup.spec.js]: remove extra line

File: test/9_UserGroup.spec.js [hunk 28]
```diff
@@ -0,0 +1,90 @@
...
+        groupAddress : erc20.address,
+        groupType: 0,
+      });
+      userGroup = await UserGroupMock.new(ERC20);
+      
+      try {
+        await userGroup.testGetAdmin();
+      } catch ({ message }) {
+        assert.strictEqual(message, getShortErrorMessage('This is not custom token group'))
+      }
```
Comment (masteralx) [inline test/9_UserGroup.spec.js]: assert.strictEqual(error, true);

File: test/9_UserGroup.spec.js [hunk 29]
```diff
@@ -0,0 +1,90 @@
...
+        groupType: 0,
+      });
+
+      userGroup = await UserGroupMock.new(nonToken);
+      try{
+        const totalSupply = await userGroup.testGetTotalSupply();
+      } catch ({message}) {
+        assert.strictEqual(message, getShortErrorMessage('').trim())
+      }
+    })
```
Comment (masteralx) [inline test/9_UserGroup.spec.js]: assert.strictEqual(error, true);

**Pull request merged**